### PR TITLE
runtime/io.c: remove the hook machinery for channel locking

### PR DESCRIPTION
**Pull request by gasche**
This is an alternative to #12444: instead of documenting hooks that are now unused by the compiler distribution, we remove them completely.

cc @xavierleroy

### CODE REVIEW

File: runtime/extern.c [hunk 1]
```diff
@@ -1094,17 +1094,17 @@ void caml_output_val(struct channel *chan, value v, value flags)
     caml_stat_free(blk);
     blk = nextblk;
   }
-  Flush_if_unbuffered(chan);
```
Comment (xavierleroy) [inline runtime/extern.c:1097]: By moving the flush to `caml_output_value`, you're changing the behavior of `caml_output_val`, which is exported (not CAML_INTERNAL) and used by the debugger.  The debugger manages its own flushing, if I remember correctly, but other potential users may not.

File: runtime/io.c [hunk 2]
```diff
@@ -77,28 +83,18 @@ static void channel_mutex_lock_default(struct channel *chan)
...
   last_channel_locked = NULL;
 }
 
-static void channel_mutex_unlock_exn_default(void)
+CAMLexport void caml_channel_cleanup_on_raise(void)
 {
   struct channel * chan = last_channel_locked;
-  if (chan != NULL && caml_channel_mutex_unlock != NULL)
-    caml_channel_mutex_unlock(chan);
+  caml_channel_unlock(chan);
```
Comment (xavierleroy) [inline runtime/io.c]: You need to test `chan != NULL` !!!

**Review by xavierleroy — COMMENTED** [review]
Something looks wrong in cleanup_on_raise.

File: runtime/io.c (hunk 2 above)
Comment (gasche) [reply runtime/io.c]: Oops! I was in a hurry to remove the other NULL test and did not pay close attention. Fixed.

File: runtime/extern.c (hunk 1 above)
Comment (gasche) [reply runtime/extern.c:1097]: The previous code looks fairly strange -- it breaks the convention of all other users of `Flush_if_unbuffered` that it happens right before we unlock, which is a helpful way to ensure that it happens often enough (no output is forgotten) and yet not too often (we are done with our planned sequence of writes).

If I read things correctly:

1. `caml_output_val` is in fact defined in intext.h only when CAML_INTERNALS is set, so we don't need to worry about unknown callers
2. this call only makes a difference for unbuffered channels, and the debugger does not use unbuffered channels (and indeed, as you mention, has an explicit flush right after the call)

This gives me the impression that the change is a no-op that makes the code more regular and less surprising. (Even after the change, `caml_output_val` can call a GC through `check_pending` in io.c, which makes the implementation of `safe_output_value` in the debugger a bit suspicious, but I don't understand what this is supposed to do and am happy to leave it alone.)

Comment (gasche) [reply runtime/extern.c:1097]: (cc @nojb who wrote the original stray `flush_if_unbuffered` call in #10639 )

Comment (xavierleroy) [reply runtime/extern.c:1097]: The mystery of the placement of this flush operation remains...  But this should not block this PR. I fixed the conflicts to get a last round of testing.

**Review by xavierleroy — APPROVED** [review]
Looks good to me,  thanks!

**Comment by gasche** [comment]
Merged, thanks!

**Comment by Octachron** [comment]
Beware that there is a conflict with the io bigarray PR which needs to be fixed.

**Comment by gasche** [comment]
On it, thanks!

(This is a non-textual conflict, so git did not detect it, but the build on trunk is currently broken.)

**Comment by gasche** [comment]
Hopefully 76c4617b90934c418fdd21b9af42bb52590754e8 fixes the build on trunk.
